docs/beamrunner/beamrunner-X-performance.md

```diff
@@ -59,6 +59,43 @@ option `--parallelWidths=4,Serial=1,VeryParallel=8` would disable
 parallelism for `Serial` while using higher parallelism for `VeryParallel`
 than the rest of the application.
 
+`PCollection`s of [`KV`](https://github.com/apache/beam/blob/release-2.4.0/sdks/java/core/src/main/java/org/apache/beam/sdk/values/KV.java) elements are partitioned based on keys. The same key always goes into
+the same parallel channel, guaranteeing that keyed states are correctly
+handled. `PCollections` of other types are partitioned using the round-robin
+scheme. It is applications' responsibility to balance workloads across keys.
+Skewed workloads can overwhelm hot keys and hamper performance. Since the
+value of keys are determined by applications, the runner cannot easily
+detect or predict keys. If a `PCollection` contains fewer keys than the
+parallel width, some parallel channels will stay idle, wasting system
+resources. One special case is a `PCollection` using `Void` key. In this
+situation, the runner can identify that there is only one global key and
+run the transform in a non-parallelized form. The following code snippet
+demonstrates an example. As the `AddKey` transform add the same `Void` key
+to all elements, the downstream `GBK` cannot run in parallel. If the

```

Comment:
After "cannot run in parallel".  Add the following: 
```
If a user specifies a parallel width (greater than 1) for the entire pipeline,
the runner may silently ignore the parallel width for the section of the pipeline 
that uses the `Void` keys. If a user explicitly specifies a parallel width for the `GBK` in the example, the runner, again, may ignore the specified width.
```
Don't know if we need to say a warning message will be printed.